File: kattis/islandhopping.cpp
#include <algorithm>
#include <cmath>
#include <fstream>
#include <iostream>
#include <queue>
#include <string>
#include <vector>

using namespace std;

struct Edge {
    int from;
    int to;
    float cost;

    Edge(int from, int to, float cost) : from(from), to(to), cost(cost) {}

    bool operator<(Edge e2) const { return cost > e2.cost; }
};

class CPSolver {
  public:
    CPSolver(bool debug);
    ~CPSolver();
    string get_string();
    int get_int();
    float get_float();
    vector<int> get_ints(int n);
    void operator()();

    float get_cost(int i, int j) {
        pair<float, float> i1 = islands[i];
        pair<float, float> i2 = islands[j];

        return sqrt(pow(i2.second - i1.second, 2) +
                    pow(i2.first - i1.first, 2));
    }

    void add_edges(int node) {
        visited[node] = true;

        for (int i = 0; i < islands.size(); i++) {
            if (visited[i])
                continue;

            pq_pointer->push(Edge(node, i, get_cost(node, i)));
        }
    }

    void solve() {
        int n = get_int();

        while (n--) {
            islands.clear();
            visited.clear();

            int m = get_int();

            for (int i = 0; i < m; i++) {
                float x = get_float();
                float y = get_float();

                islands.push_back(make_pair(x, y));
                visited.push_back(false);
            }

            priority_queue<Edge> pq;
            pq_pointer = &pq;
            add_edges(0);

            float total_cost = 0;
            int mst_nodes = 1;

            while (pq.size() != 0 && mst_nodes < m) {
                Edge e = pq.top();
                pq.pop();

                if (visited[e.to])
                    continue;

                mst_nodes += 1;
                total_cost += e.cost;
                add_edges(e.to);
            }

            cout << total_cost << endl;
        }
    }

  private:
    vector<pair<float, float>> islands;
    vector<bool> visited;
    priority_queue<Edge> *pq_pointer;

  private:
    bool debug;
    istream *inp;
    ifstream *finp;
};

int main() {
    CPSolver solver(true);
    solver();

    return 0;
}

CPSolver::CPSolver(bool debug) {
    this->debug = debug;
    if (debug) {
        finp = new ifstream();
        finp->open("../input.txt");
        inp = finp;
    } else {
        inp = &cin;
    }
}

CPSolver::~CPSolver() {
    if (debug) {
        finp->close();
    }
}

void CPSolver::operator()() { solve(); }

string CPSolver::get_string() {
    string str;
    getline(*inp, str);
    return str;
}

int CPSolver::get_int() {
    int n;
    (*inp) >> n;
    return n;
}

vector<int> CPSolver::get_ints(int n) {
    if (n <= 0)
        return {};
    vector<int> ret;
    while (n--) {
        ret.push_back(get_int());
    }
    return ret;
}

float CPSolver::get_float() {
    float n;
    (*inp) >> n;
    return n;
}

File: kattis/islandhopping.py
"""
Link:
"""

import sys
import heapq
import math

DEBUG = True

input_file = None
if DEBUG:
    input_file = open("../input.txt", "r", encoding="utf-8")


def get_string() -> str:
    """Returns a string from the input."""

    if DEBUG and input_file is not None:
        line = input_file.readline()
    else:
        line = sys.stdin.readline()
    return line.strip()


def get_int() -> int:
    """Returns an integer from the input."""

    return int(get_string())


def get_ints() -> list[int]:
    """Returns integers from the input."""

    return list(map(int, get_string().split()))


def get_float() -> float:
    """Returns a float from the input."""

    return float(get_string())


def get_floats() -> list[float]:
    """Returns floats from the input."""

    return list(map(float, get_string().split()))


def calc_distance(nodes, i, j) -> float:
    x1, y1 = nodes[i]
    x2, y2 = nodes[j]

    return math.sqrt((x2 - x1) ** 2 + (y2 - y1) ** 2)


def add_edges(visited, pq, nodes, node):
    visited[node] = True

    for i in range(len(nodes)):
        if visited[i]:
            continue

        heapq.heappush(pq, (calc_distance(nodes, node, i), node, i))


def main():
    n = get_int()
    for _ in range(n):

        m = get_int()
        islands = []
        for _ in range(m):
            x, y = get_floats()
            islands.append((x, y))

        # Prim's algo
        mst_cost = 0.0
        mst_node_count = 1
        visited = [False] * m

        pq = []
        add_edges(visited, pq, islands, 0)
        while len(pq) != 0 and mst_node_count < m:

            cost, _, end = heapq.heappop(pq)

            if visited[end]:
                continue

            add_edges(visited, pq, islands, end)
            mst_node_count += 1
            mst_cost += cost

        print(mst_cost)


if __name__ == "__main__":
    main()
